spikeextractors/extractors/neuroscopeextractors/neuroscopeextractors.py

```diff
@@ -0,0 +1,387 @@
+            any_res = any(end_res)
+            any_clu = any(end_clu)
+            
+            if any_res or any_clu:
+                assert any_res == True and any_clu == True, 'Unmatched .res and .clu files detected!'
+                
+                resfile_path = '{}/{}.res'.format(folder_path,SORTING_NAME)
+                clufile_path = '{}/{}.clu'.format(folder_path,SORTING_NAME)
+                # NeuroscopeSortingExtractor(self,resfile_path=resfile_path,
+                #                            clufile_path=clufile_path,
+                #                            keep_mua_units=keep_mua_units)
+                NeuroscopeSortingExtractor.__init__(self, resfile_path=resfile_path,
+                                                          clufile_path=clufile_path,
+                                                          keep_mua_units=keep_mua_units)
+            else:
+                shank_res_ids = [x[-1] for x in onlyfiles if x[-5:-2] == 'res' and x[-10:-6] != 'temp']
+                shank_clu_ids = [x[-1] for x in onlyfiles if x[-5:-2] == 'clu' and x[-10:-6] != 'temp']
+
+                assert shank_res_ids==shank_clu_ids, 'Unmatched .clu.%i and .res.%i files detected!'
+                
+                all_shanks_list_se = []
+                for shank_id in shank_res_ids:
+                    resfile_path = '{}/{}.res.{}'.format(folder_path,SORTING_NAME,shank_id)
+                    clufile_path = '{}/{}.clu.{}'.format(folder_path,SORTING_NAME,shank_id)
+                    
+                    
+                    # when called with either NeuroscopeSortingExtractor(MultiSortingExtractor) or NeuroscopeSortingExtractor(SortingExtractor)
+                    # doesn't recognize the items in the sorting as being NeuroscopeSortingExtractors, thinks they are 'Nonetype'
+                    all_shanks_list_se.append(NeuroscopeSortingExtractor.__init__(self, resfile_path=resfile_path,
+                                                                                        clufile_path=clufile_path,
+                                                                                        keep_mua_units=keep_mua_units))
+                    
+                    # runs correct, but final object does not inherit basic functions like 'get_unit_ids' from MultiSortingExtractor
+                    # and also doesn't seem to set the interval values of _all_unit_ids and such correctly...
+                    # Maybe the issue lies in calling NeuroscopeSortingExtractor() within its own __init__?
+                    # all_shanks_list_se.append(NeuroscopeSortingExtractor(resfile_path=resfile_path,
+                    #                                                      clufile_path=clufile_path,
+                    #                                                      keep_mua_units=keep_mua_units))
+                    
+                    
+                MultiSortingExtractor.__init__(self, all_shanks_list_se)
```

Comment:
I think that what you are missing is the following:

`self = MultiSortingExtractor.__init__(self, all_shanks_list_se)`

This basically resets the object to be a `MultiSortingExtractor`, hence inheriting all the `SortingExtractor` methods.

The same holds for line 235:
`self = NeuroscopeSortingExtractor.__init__(self, resfile_path=resfile_path,
                                                          clufile_path=clufile_path,
                                                          keep_mua_units=keep_mua_units)`

I think this could also be a viable solution in order not to have 2 separate extractors as @CodyCBakerPhD suggested

Reply:
Moreover, @CodyCBakerPhD you sould set the group property for each unit: at line 252:

```
single_sorting = NeuroscopeSortingExtractor(resfile_path=resfile_path, clufile_path=clufile_path, 
                                                                              keep_mua_units=keep_mua_units)
for u in single_sorting.get_unit_ids():
    single_sorting.set_unit_property(u, 'group', int(shank_id))
all_shanks_list_se.append(single_sorting)
```

So when you create the `MultiSortingExtractor`, each unit carries info about its group :)

Reply:
In the end it doesn't seem to matter if I do `self = MultiSortingExtractor(sortings=all_shanks_list_se)`  or `MultiSortingExtractor.__init__(self,sortings=all_shanks_list_se)` to create and fill that object; the issue is apparently the dual overriding of the `get_unit_ids()` function, which calls different internal names for the two base types (`_unit_ids` or `_all_unit_ids`). If the entire NeuroscopeSortingExtractor class inherits from SortingExtractor, everything works during initialization, and for single shank conditions, but for multi-shank the `get_unit_ids()` calls the function specific to NeuroscopeSortingExtractors instead of the MultiSortingExtractor, even if we set `self = MultiSortingExtractor(sortings=all_shanks_list_se)`. Likewise, if we inherit from the MultiSortingExtractor and allow it to override that function definition, then the single shank versions stop working since they use the different internal naming conventions. 

Similar issues to this would also arise for `get_unit_spike_train()`. 

I can only see two fixes to this: 

1) Adapt the `get_unit_ids()` function for the NeuroscopeSortingExtractor class to identify whether or not it is a single shank or multi shank and call the correct corresponding id retrieval method. Ditto for spike trains.

2) Two separate classes, with the NeuroscopeMultiSortingExtractor using thin inheritance, and the NeuroscopeSortingExtractor being exactly what it is now without the multi-shank condition (which is all working as intended).

I'm inclined to go with solution 2 here since it keeps the application of the methods abstract and appropriate to their final use. This can then be refactored later if a better base approach to multiple data sources is incorporated into the extractors.

Reply:
Hi @CodyCBakerPhD 

I though it would have worked! I also prefer solution 2 then. The user would now if there are multiple shanks or not, or at least this is easily discoverable by looking at the folder. 

Thanks for looking into this :)